### ISSUE TITLE
Qdb "step" option cause a crash (Unicorn API has changed)

### ISSUE THREAD

**Issue by ciapatakoza** [qilingframework/qiling#718]
The Qdb "step" and "p" options leads to the following crash:

```
  File "hello_world_linux.py", line 92, in <module>
    my_sandbox(["/home/vm/Software/linux_rootfs/bin/hello_app"], "/home/vm/Software/linux_rootfs/")
  File "hello_world_linux.py", line 76, in my_sandbox
    ql.run()
  File "/home/vm/.local/lib/python3.8/site-packages/qiling/core.py", line 756, in run
    self.os.run()
  File "/home/vm/.local/lib/python3.8/site-packages/qiling/os/linux/linux.py", line 90, in run
    self.ql.emu_start(self.ql.loader.entry_point, self.ql.loader.elf_entry, self.ql.timeout)
  File "/home/vm/.local/lib/python3.8/site-packages/qiling/core.py", line 900, in emu_start
    raise self._internal_exception
  File "/home/vm/.local/lib/python3.8/site-packages/qiling/utils.py", line 156, in wrapper
    return func(*args, **kw)
  File "/home/vm/.local/lib/python3.8/site-packages/qiling/core_hooks.py", line 210, in _hook_addr_cb
    ret = h.call(ql, addr, size)
  File "/home/vm/.local/lib/python3.8/site-packages/qiling/core_hooks.py", line 46, in call
    return self.callback(ql)
  File "/home/vm/.local/lib/python3.8/site-packages/qiling/debugger/qdb/qdb.py", line 112, in _breakpoint_handler
    self.interactive()
  File "/home/vm/.local/lib/python3.8/site-packages/qiling/debugger/qdb/qdb.py", line 59, in interactive
    return self.cmdloop()
  File "/usr/lib/python3.8/cmd.py", line 138, in cmdloop
    stop = self.onecmd(line)
  File "/usr/lib/python3.8/cmd.py", line 217, in onecmd
    return func(arg)
  File "/home/vm/.local/lib/python3.8/site-packages/qiling/debugger/qdb/qdb.py", line 172, in do_step
    self._states_list.append(diff_snapshot_save(current_state_dicts, self._states_list[-1]))
  File "/home/vm/.local/lib/python3.8/site-packages/qiling/debugger/qdb/utils.py", line 420, in diff_snapshot_save
    cur_cpu_ctx_set = {(idx, val) for idx, val in enumerate(current_state_dicts["cpu_context"])}
TypeError: 'UcContext' object is not iterable
```

I had a look on this and it seems that Unicron API has changed and now the UcContext object is returned from the:
`self.ql.uc.context_save()`

**Comment by wtdcode** [comment]
Seems so. Will have a look tomorrow.

________________________________
From: ciapatakoza ***@***.***>
Sent: Thursday, March 11, 2021 10:31:58 PM
To: qilingframework/qiling ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [qilingframework/qiling] Qdb "step" option cause a crash (Unicorn API has changed) (#718)


The Qdb "step" and "p" options leads to the following crash:
File "hello_world_linux.py", line 92, in <module> my_sandbox(["/home/vm/Software/linux_rootfs/bin/camera_app"], "/home/vm/Software/linux_rootfs/") File "hello_world_linux.py", line 76, in my_sandbox ql.run() File "/home/vm/.local/lib/python3.8/site-packages/qiling/core.py", line 756, in run self.os.run() File "/home/vm/.local/lib/python3.8/site-packages/qiling/os/linux/linux.py", line 90, in run self.ql.emu_start(self.ql.loader.entry_point, self.ql.loader.elf_entry, self.ql.timeout) File "/home/vm/.local/lib/python3.8/site-packages/qiling/core.py", line 900, in emu_start raise self._internal_exception File "/home/vm/.local/lib/python3.8/site-packages/qiling/utils.py", line 156, in wrapper return func(*args, **kw) File "/home/vm/.local/lib/python3.8/site-packages/qiling/core_hooks.py", line 210, in _hook_addr_cb ret = h.call(ql, addr, size) File "/home/vm/.local/lib/python3.8/site-packages/qiling/core_hooks.py", line 46, in call return self.callback(ql) File "/home/vm/.local/lib/python3.8/site-packages/qiling/debugger/qdb/qdb.py", line 112, in _breakpoint_handler self.interactive() File "/home/vm/.local/lib/python3.8/site-packages/qiling/debugger/qdb/qdb.py", line 59, in interactive return self.cmdloop() File "/usr/lib/python3.8/cmd.py", line 138, in cmdloop stop = self.onecmd(line) File "/usr/lib/python3.8/cmd.py", line 217, in onecmd return func(arg) File "/home/vm/.local/lib/python3.8/site-packages/qiling/debugger/qdb/qdb.py", line 172, in do_step self._states_list.append(diff_snapshot_save(current_state_dicts, self._states_list[-1])) File "/home/vm/.local/lib/python3.8/site-packages/qiling/debugger/qdb/utils.py", line 420, in diff_snapshot_save cur_cpu_ctx_set = {(idx, val) for idx, val in enumerate(current_state_dicts["cpu_context"])} TypeError: 'UcContext' object is not iterable

I had a look on this and it seems that Unicron API has changed and now the UcContext object is returned from the:
self.ql.uc.context_save()

―

**Comment by wtdcode** [comment]
What's your qiling and unicorn version? Seems not to be reproducible on my side.

**Comment by wtdcode** [comment]
Also cc @ucgJhe

**Comment by ciapatakoza** [comment]
```
$ pip3 list | grep -E "unicorn|qiling"
qiling                 1.2.2
unicorn                1.0.2
```

**Comment by ucgJhe** [comment]
> ```
> $ pip3 list | grep -E "unicorn|qiling"
> qiling                 1.2.2
> unicorn                1.0.2
> ```

hi, qdb is unable to do diff snapshot for saving/restoring CPU state due to unicorn API update, i will push PR to fix it.

**Comment by wtdcode** [comment]
Fixed in #746.
